cibuildwheel GitHub action fails due to duplicate skbuild build directory on Windows ARM64 build

### Description

I hit a build failure when using skbuild in conjunction with cibuildwheel. 

Here is the relevant part of the build log:

```console
     Configuring Project
      Working directory:
        D:\a\micropython-uncrustify\micropython-uncrustify\_skbuild\win-amd64-3.11\cmake-build
      Command:
        'C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-x7dj25c1\overlay\Lib\site-packages\cmake\data\bin/cmake.exe' 'D:\a\micropython-uncrustify\micropython-uncrustify' -G 'Visual Studio 17 2022' --no-warn-unused-cli '-DCMAKE_INSTALL_PREFIX:PATH=D:\a\micropython-uncrustify\micropython-uncrustify\_skbuild\win-amd64-3.11\cmake-install' -DPYTHON_VERSION_STRING:STRING=3.11.9 -DSKBUILD:INTERNAL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-x7dj25c1\overlay\Lib\site-packages\skbuild\resources\cmake' '-DPYTHON_EXECUTABLE:PATH=C:\Users\runneradmin\AppData\Local\Temp\cibw-run-dk_dqzxv\cp311-win_arm64\build\venv\Scripts\python.exe' '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.11.9\tools\Include' '-DPYTHON_LIBRARY:PATH=C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\pythonarm64.3.11.9\tools\libs\python311.lib' '-DPython_EXECUTABLE:PATH=C:\Users\runneradmin\AppData\Local\Temp\cibw-run-dk_dqzxv\cp311-win_arm64\build\venv\Scripts\python.exe' '-DPython_ROOT_DIR:PATH=C:\Users\runneradmin\AppData\Local\Temp\cibw-run-dk_dqzxv\cp311-win_arm64\build\venv' -DPython_FIND_REGISTRY:STRING=NEVER '-DPython_INCLUDE_DIR:PATH=C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.11.9\tools\Include' '-DPython_LIBRARY:PATH=C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\pythonarm64.3.11.9\tools\libs\python311.lib' '-DPython3_EXECUTABLE:PATH=C:\Users\runneradmin\AppData\Local\Temp\cibw-run-dk_dqzxv\cp311-win_arm64\build\venv\Scripts\python.exe' '-DPython3_ROOT_DIR:PATH=C:\Users\runneradmin\AppData\Local\Temp\cibw-run-dk_dqzxv\cp311-win_arm64\build\venv' -DPython3_FIND_REGISTRY:STRING=NEVER '-DPython3_INCLUDE_DIR:PATH=C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.11.9\tools\Include' '-DPython3_LIBRARY:PATH=C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\pythonarm64.3.11.9\tools\libs\python311.lib' -T v143 -A ARM64 -DCMAKE_BUILD_TYPE:STRING=Release
    
    Not searching for unused variables given on the command line.
    CMake Error: Error: generator platform: ARM64
    Does not match the platform used previously: x64
    Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory.
    Traceback (most recent call last):
      File "C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-x7dj25c1\overlay\Lib\site-packages\skbuild\setuptools_wrap.py", line 666, in setup
        env = cmkr.configure(
              ^^^^^^^^^^^^^^^
      File "C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-x7dj25c1\overlay\Lib\site-packages\skbuild\cmaker.py", line 357, in configure
        raise SKBuildError(msg)
```

We can see that this is targeting ARM64 Windows, but the build directory is `_skbuild\win-amd64-3.11\cmake-build`. This is the same build directory that was already used by the amd64 build in the same CI job. So cmake is giving the expected error.

I did some digging into the code and this is happening because skbuild is `distutils.util.get_platform()` to [get the platform](https://github.com/scikit-build/scikit-build/blob/4dab4576d7a480da7484cfc5c249c86f7d3ecde3/skbuild/constants.py#L47), so it is getting the host platform instead of the target platform.

This can be [overridden by setting the VSCMD_ARG_TGT_ARCH environment variable](https://github.com/pypa/setuptools/blob/f91fa3d9fc7262e0467e4b2f84fe463f8f8d23cf/setuptools/_distutils/util.py#L52).

So I added the following to my `pyproject.toml`:

```toml
# HACK: causes distutils to return the proper platform name so that scikit-build
# will not use the same build directory as the host platform build.
[[tool.cibuildwheel.overrides]]
select = "*-win_arm64"
environment = "VSCMD_ARG_TGT_ARCH=arm64"
```

It would be nice if this "just worked" or was at least the workaround was documented in the various examples.

FWIW, using `VSCMD_ARG_TGT_ARCH` was also mentioned in https://github.com/pypa/cibuildwheel/pull/1144#issuecomment-1347189520



### Build log

https://github.com/dlech/micropython-uncrustify/actions/runs/9429791938/job/25976561751

### CI config

https://github.com/dlech/micropython-uncrustify/actions/runs/9429791938/workflow

--- a/cibuildwheel/windows.py
+++ b/cibuildwheel/windows.py
@@ -157,8 +157,8 @@ def setup_setuptools_cross_compile(
     # identifies the target, not the host
     vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
     current_tgt_arch = vscmd_arg_tgt_arch[python_configuration.arch]
-    if env.get("VSCMD_ARG_TGT_ARCH", current_tgt_arch) != current_tgt_arch:
-        msg = f"VSCMD_ARG_TGT_ARCH must be set to {current_tgt_arch}, got {env['VSCMD_ARG_TGT_ARCH']}. Make sure you setup MSVC targeting the right architecture."
+    if (env.get("VSCMD_ARG_TGT_ARCH") or current_tgt_arch) != current_tgt_arch:
+        msg = f"VSCMD_ARG_TGT_ARCH must be set to {current_tgt_arch!r}, got {env['VSCMD_ARG_TGT_ARCH']!r}. If you're setting up MSVC yourself (e.g. using vcvarsall.bat or msvc-dev-cmd), make sure to target the right architecture. Alternatively, run cibuildwheel without configuring MSVC, and let the build backend handle it."
         raise errors.FatalError(msg)
     env["VSCMD_ARG_TGT_ARCH"] = current_tgt_arch
 

--- a/cibuildwheel/windows.py
+++ b/cibuildwheel/windows.py
@@ -155,8 +155,9 @@ def setup_setuptools_cross_compile(
 
     # Set environment variable so that setuptools._distutils.get_platform()
     # identifies the target, not the host
-    vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
-    env["VSCMD_ARG_TGT_ARCH"] = vscmd_arg_tgt_arch[python_configuration.arch]
+    if not env.get("VSCMD_ARG_TGT_ARCH"):
+        vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
+        env["VSCMD_ARG_TGT_ARCH"] = vscmd_arg_tgt_arch[python_configuration.arch]
 
     # (This file must be default/locale encoding, so we can't pass 'encoding')
     distutils_cfg.write_text(

--- a/cibuildwheel/windows.py
+++ b/cibuildwheel/windows.py
@@ -155,9 +155,12 @@ def setup_setuptools_cross_compile(
 
     # Set environment variable so that setuptools._distutils.get_platform()
     # identifies the target, not the host
-    if not env.get("VSCMD_ARG_TGT_ARCH"):
-        vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
-        env["VSCMD_ARG_TGT_ARCH"] = vscmd_arg_tgt_arch[python_configuration.arch]
+    vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
+    current_tgt_arch = vscmd_arg_tgt_arch[python_configuration.arch]
+    if env.get("VSCMD_ARG_TGT_ARCH", current_tgt_arch) != current_tgt_arch:
+         msg = f"VSCMD_ARG_TGT_ARCH must be set to {current_tgt_arch}, got {env['VSCMD_ARG_TGT_ARCH']}. Make sure you setup MSVC targetting the right architecture."
+         raise errors.FatalError(msg)
+    env["VSCMD_ARG_TGT_ARCH"] = current_tgt_arch
 
     # (This file must be default/locale encoding, so we can't pass 'encoding')
     distutils_cfg.write_text(

--- a/cibuildwheel/windows.py
+++ b/cibuildwheel/windows.py
@@ -158,8 +158,8 @@ def setup_setuptools_cross_compile(
     vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
     current_tgt_arch = vscmd_arg_tgt_arch[python_configuration.arch]
     if env.get("VSCMD_ARG_TGT_ARCH", current_tgt_arch) != current_tgt_arch:
-         msg = f"VSCMD_ARG_TGT_ARCH must be set to {current_tgt_arch}, got {env['VSCMD_ARG_TGT_ARCH']}. Make sure you setup MSVC targetting the right architecture."
-         raise errors.FatalError(msg)
+        msg = f"VSCMD_ARG_TGT_ARCH must be set to {current_tgt_arch}, got {env['VSCMD_ARG_TGT_ARCH']}. Make sure you setup MSVC targeting the right architecture."
+        raise errors.FatalError(msg)
     env["VSCMD_ARG_TGT_ARCH"] = current_tgt_arch
 
     # (This file must be default/locale encoding, so we can't pass 'encoding')

--- a/cibuildwheel/windows.py
+++ b/cibuildwheel/windows.py
@@ -152,6 +152,12 @@ def setup_setuptools_cross_compile(
     # set the platform name
     map_plat = {"32": "win32", "64": "win-amd64", "ARM64": "win-arm64"}
     plat_name = map_plat[python_configuration.arch]
+
+    # Set environment variable so that setuptools._distutils.get_platform()
+    # identifies the target, not the host
+    vscmd_arg_tgt_arch = {"32": "x86", "64": "x64", "ARM64": "arm64"}
+    env["VSCMD_ARG_TGT_ARCH"] = vscmd_arg_tgt_arch[python_configuration.arch]
+
     # (This file must be default/locale encoding, so we can't pass 'encoding')
     distutils_cfg.write_text(
         textwrap.dedent(

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -75,6 +75,7 @@ test = [
     "pytest-timeout",
     "pytest-xdist",
     "pytest>=6",
+    "setuptools",
     "tomli_w",
     "validate-pyproject",
 ]
@@ -140,6 +141,8 @@ disallow_untyped_decorators = true
 [[tool.mypy.overrides]]
 module = [
     "setuptools",
+    "setuptools._distutils", # needed even if only directly import setuptools._distutils.util
+    "setuptools._distutils.util",
     "pytest", # ignored in pre-commit to speed up check
     "bashlex",
     "bashlex.*",

--- a/unit_test/get_platform_test.py
+++ b/unit_test/get_platform_test.py
@@ -7,6 +7,7 @@ from typing import Dict
 import pytest
 import setuptools._distutils.util
 
+from cibuildwheel.util import CIProvider, detect_ci_provider
 from cibuildwheel.windows import PythonConfiguration, setup_setuptools_cross_compile
 
 # monkeypatching os.name is too flaky. E.g. It works on my machine, but fails in pipeline
@@ -50,6 +51,9 @@ def test_x64(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
     assert target_platform == "win-amd64"
 
 
+@pytest.mark.skipif(
+    detect_ci_provider() == CIProvider.azure_pipelines, reason="arm64 not recognised on azure"
+)
 def test_arm(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
     arch = "ARM64"
     environment: Dict[str, str] = {}
@@ -66,7 +70,7 @@ def test_arm(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
 
 def test_env_set(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
     arch = "32"
-    environment = {"VSCMD_ARG_TGT_ARCH": "arm64"}
+    environment = {"VSCMD_ARG_TGT_ARCH": "x64"}
 
     configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
 
@@ -74,8 +78,8 @@ def test_env_set(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
     with patched_environment(monkeypatch, environment):
         target_platform = setuptools._distutils.util.get_platform()
 
-    assert environment["VSCMD_ARG_TGT_ARCH"] == "arm64"
-    assert target_platform == "win-arm64"
+    assert environment["VSCMD_ARG_TGT_ARCH"] == "x64"
+    assert target_platform == "win-amd64"
 
 
 def test_env_blank(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):

--- a/unit_test/get_platform_test.py
+++ b/unit_test/get_platform_test.py
@@ -1,5 +1,6 @@
 # ruff: noqa: ARG001
 import contextlib
+import sys
 from pathlib import Path
 from typing import Dict
 
@@ -8,11 +9,14 @@ import setuptools._distutils.util
 
 from cibuildwheel.windows import PythonConfiguration, setup_setuptools_cross_compile
 
+# monkeypatching os.name is too flaky. E.g. It works on my machine, but fails in pipeline
+if not sys.platform.startswith("win"):
+    pytest.skip("Windows-only tests", allow_module_level=True)
+
 
 @contextlib.contextmanager
 def patched_environment(monkeypatch: pytest.MonkeyPatch, environment: Dict[str, str]):
     with monkeypatch.context() as mp:
-        mp.setattr("os.name", "nt")
         for envvar, val in environment.items():
             mp.setenv(name=envvar, value=val)
         yield

--- a/unit_test/get_platform_test.py
+++ b/unit_test/get_platform_test.py
@@ -1,0 +1,88 @@
+# ruff: noqa: ARG001
+import contextlib
+from pathlib import Path
+from typing import Dict
+
+import pytest
+import setuptools._distutils.util
+
+from cibuildwheel.windows import PythonConfiguration, setup_setuptools_cross_compile
+
+
+@contextlib.contextmanager
+def patched_environment(monkeypatch: pytest.MonkeyPatch, environment: Dict[str, str]):
+    with monkeypatch.context() as mp:
+        mp.setattr("os.name", "nt")
+        for envvar, val in environment.items():
+            mp.setenv(name=envvar, value=val)
+        yield
+
+
+def test_x86(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    arch = "32"
+    environment: Dict[str, str] = {}
+
+    configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
+
+    setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
+    with patched_environment(monkeypatch, environment):
+        target_platform = setuptools._distutils.util.get_platform()
+
+    assert environment["VSCMD_ARG_TGT_ARCH"] == "x86"
+    assert target_platform == "win32"
+
+
+def test_x64(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    arch = "64"
+    environment: Dict[str, str] = {}
+
+    configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
+
+    setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
+    with patched_environment(monkeypatch, environment):
+        target_platform = setuptools._distutils.util.get_platform()
+
+    assert environment["VSCMD_ARG_TGT_ARCH"] == "x64"
+    assert target_platform == "win-amd64"
+
+
+def test_arm(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    arch = "ARM64"
+    environment: Dict[str, str] = {}
+
+    configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
+
+    setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
+    with patched_environment(monkeypatch, environment):
+        target_platform = setuptools._distutils.util.get_platform()
+
+    assert environment["VSCMD_ARG_TGT_ARCH"] == "arm64"
+    assert target_platform == "win-arm64"
+
+
+def test_env_set(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    arch = "32"
+    environment = {"VSCMD_ARG_TGT_ARCH": "arm64"}
+
+    configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
+
+    setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
+    with patched_environment(monkeypatch, environment):
+        target_platform = setuptools._distutils.util.get_platform()
+
+    assert environment["VSCMD_ARG_TGT_ARCH"] == "arm64"
+    assert target_platform == "win-arm64"
+
+
+def test_env_blank(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+    arch = "32"
+    environment = {"VSCMD_ARG_TGT_ARCH": ""}
+
+    configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
+
+    setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
+    with patched_environment(monkeypatch, environment):
+        target_platform = setuptools._distutils.util.get_platform()
+
+    assert environment["VSCMD_ARG_TGT_ARCH"] == "x86"
+    assert target_platform == "win32"

--- a/unit_test/get_platform_test.py
+++ b/unit_test/get_platform_test.py
@@ -1,4 +1,3 @@
-# ruff: noqa: ARG001
 import contextlib
 import sys
 from pathlib import Path

--- a/unit_test/get_platform_test.py
+++ b/unit_test/get_platform_test.py
@@ -7,6 +7,7 @@ from typing import Dict
 import pytest
 import setuptools._distutils.util
 
+from cibuildwheel.errors import FatalError
 from cibuildwheel.util import CIProvider, detect_ci_provider
 from cibuildwheel.windows import PythonConfiguration, setup_setuptools_cross_compile
 
@@ -68,18 +69,14 @@ def test_arm(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
     assert target_platform == "win-arm64"
 
 
-def test_env_set(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):
+def test_env_set(tmp_path: Path):
     arch = "32"
     environment = {"VSCMD_ARG_TGT_ARCH": "x64"}
 
     configuration = PythonConfiguration("irrelevant", arch, "irrelevant", None)
 
-    setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
-    with patched_environment(monkeypatch, environment):
-        target_platform = setuptools._distutils.util.get_platform()
-
-    assert environment["VSCMD_ARG_TGT_ARCH"] == "x64"
-    assert target_platform == "win-amd64"
+    with pytest.raises(FatalError, match="VSCMD_ARG_TGT_ARCH"):
+        setup_setuptools_cross_compile(tmp_path, configuration, tmp_path, environment)
 
 
 def test_env_blank(tmp_path: Path, monkeypatch: pytest.MonkeyPatch):

